syntax = "proto3";

package kvStore;

// services
service KvNodeService {
    rpc SayHello (HelloRequest) returns (HelloReply) {}

    rpc Request(RequestContent) returns (RequestResult) {}
    rpc Sync(SyncContent) returns (SyncResult) {}
}

// hello messages
message HelloRequest {
    string name = 1;
}

message HelloReply { string message = 1; }

// kvstore communication messages
message RequestContent {
  string key = 1;
  string value = 2;
  int64 op = 3;
}

message RequestResult {
  string value = 1;
  int64 err = 2;
}

// sync messages
message SyncContent {
  int64 pterm = 1;
  int64 pindex = 2;
  int64 term = 3;
  int64 index = 4;
  RequestContent req = 5;
}

message SyncResult {
  int64 err = 1;
}